[1.20.1] Added TPS activate event, modified BuildCameraTransform

Yesssssman opened the pull request:
Added a new event to control the vanilla camera process

In EpicFightCameraAPI, line 793
```java
public boolean setupCamera(Camera camera, float partialTick) {
		if (!camera.getEntity().is(this.minecraft.player)) {
			return false;
		}
		
		BuildCameraTransform.Pre event = new BuildCameraTransform.Pre(this, camera, partialTick);
		EpicFightClientHooks.Camera.BUILD_TRANSFORM_PRE.post(event);
		
		if (event.hasCanceled()) {
			return event.cancelVanillaCameraSetup(); // Exposed the event to control the return
		}
...
```

This allows the event to completely turn off the Epic Fight modification and let Minecraft handle the camera setup. The method return value is used by `MixinCamera`

```java
public void epicfight$setup(BlockGetter level, Entity entity, boolean detached, boolean thirdPersonReverse, float partialTick, CallbackInfo callbackInfo) {
	this.initialized = true;
	this.level = level;
	this.entity = entity;
	this.detached = detached;
	
	EpicFightCameraAPI cameraApi = EpicFightCameraAPI.getInstance();
	Camera camera = (Camera)(Object)this;
	
	if (cameraApi.setupCamera(camera, partialTick)) { // where `setupCamera` is called
		callbackInfo.cancel();
	}
	
	cameraApi.fireCameraBuildPost(camera, partialTick);
}
```

Additionally, a new event to control TPS camera activation has been introduced. `IsTPSMode` is widely used to check Epic Fight camera modification, so this event allows turning off all TPS camera features at once.

To demonstrate the action, I created an event hook that should disable TPS camera when the player is using an item. Without the new activation event, this was our trial to disable Epic Fight TPS (with Shoulder Surfing Reloaded, by @EchoEllet)

```java
EpicFightClientHooks.Camera.BUILD_TRANSFORM_PRE.registerEvent(event -> {
        if (Minecraft.getInstance().player != null && Minecraft.getInstance().player.isUsingItem()) {
        	event.setCancelVanillaCameraSetup(false);
        	event.cancel();
        }
});
```

This might lead to misfunctionality since `IsTPSMode()` still returns true and the mouse input is still being interrupted.

https://github.com/user-attachments/assets/594d05c1-643c-459d-9865-d2473055e9e3

The new event completely solved this issue.

```java
EpicFightClientHooks.Camera.ACTIVATE_TPS_CAMERA.registerEvent(event -> {
    if (Minecraft.getInstance().player != null && Minecraft.getInstance().player.isUsingItem()) {
        event.cancel();
    }
});
```

https://github.com/user-attachments/assets/5aae4fbd-2b27-4055-b88a-1037b0f9c0f1

Yesssssman committed:
Added TPS activate event, modified BuildCameraTransform

Yesssssman committed:
simply a name fix

Yesssssman committed:
update communication_helpr.dll corresponding to #2264

EchoEllet commented on src/main/java/yesman/epicfight/api/client/event/types/BuildCameraTransform.java:
```diff
@@ -23,9 +23,26 @@ public float getPartialTick() {
+		 * and don't want it to be overwrite by vanilla camera setups.
+		 * 
+		 * @see MixinCamera#epicfight$setup
+		 */
```
```suggestion
		/// Set to `true` to disable the vanilla camera setup process.
		/// Indicates that the camera transform has been modified through this event
		/// and should not be overwritten by vanilla camera setups.
		///
		/// @see MixinCamera#epicfight$setup
```

We should keep any new comments in Markdown to ease the migration in the future.

I have also made some other changes to keep the doc comments slightly easier to read and formal, without saying `You need to` or `we have to` or `but this could`.

Yesssssman replied:
For now, I just kept old comment style in 1.20.1 since I use eclipse and others all use javadoc style comments.

I uses markdown standard only in 1.21.1.

EchoEllet replied:
> For now, I just kept old comment style in 1.20.1 since I use eclipse and others all use javadoc style comments.

I think that's fine, the doc comment is still readable, and we should not wait for the current JDK to be exactly `23` to make it easier to handle the change.

Does the doc comment not get displayed in Eclipse?

EchoEllet commented on src/main/java/yesman/epicfight/api/event/EventHook.java:
```diff
@@ -45,31 +45,31 @@ public boolean post(T eventInstance) {
 	 */
-	public void registerPassiveEvent(DefaultEventSubscription<T> subscription, int priority) {
-		this.registerPassiveEvent(subscription, getDefaultSubscriberName(), priority);
+	public void registerEvent(DefaultEventSubscription<T> subscription, int priority) {
+		this.registerEvent(subscription, getDefaultSubscriberName(), priority);
```
Good change!
To be honest, I was hesitant about `registerPassiveEvent`

Yesssssman committed:
Update src/main/java/yesman/epicfight/api/client/event/types/BuildCameraTransform.java

Co-authored-by: Ellet <echo.ellet@gmail.com>

EchoEllet reviewed (approved):
LGTM, still need more testing with Shoulder Surfing though

EchoEllet commented on src/main/java/yesman/epicfight/api/client/event/types/BuildCameraTransform.java:
```diff
@@ -23,9 +23,24 @@ public float getPartialTick() {
+		///
+		/// @see MixinCamera#epicfight$setup
+		public void setCancelVanillaCameraSetup(boolean flag) {
+			this.cancelVanillaCameraSetup = flag;
+		}
+		
+		public boolean cancelVanillaCameraSetup() {
+			return this.cancelVanillaCameraSetup;
+		}
 	}
 	
```
I think `cancelVanillaCameraSetup` can be confused as if it has a side effect or cancels something, but it just returns `cancelVanillaCameraSetup`.

<img width="1240" height="392" alt="Image" src="https://github.com/user-attachments/assets/6babfc35-ad41-4f25-b960-faa9fc65f7ce" />

My suggestion is to refactor to `isCancelVanillaCameraSetup` or any other name.

Yesssssman committed:
Modified BuildCameraTransform.Pre event

Fixed BuildCameraTransform.Pre not to cancel the vanilla setup process
if it didn't call Camera#setPosition or Camera#setRotation.

Yesssssman committed:
Merge remote-tracking branch
'origin/chore/1.20.1-camera-api-advancements' into
chore/1.20.1-camera-api-advancements

Yesssssman committed:
Merge branch '1.20.1' into chore/1.20.1-camera-api-advancements

Yesssssman commented:
NOTE: This branch will be used to backport SSR compatibility events from 1.21.1
